File: backend/src/main/java/com/gxs/bank/controller/FixedDepositController.java
package com.gxs.bank.controller;

import com.gxs.bank.dto.response.ApiResponse;
import com.gxs.bank.model.FixedDeposit;
import com.gxs.bank.model.SavingsAccount;
import com.gxs.bank.model.User;
import com.gxs.bank.service.FixedDepositService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/fd")
@RequiredArgsConstructor
public class FixedDepositController {
    private final FixedDepositService fixedDepositService;

    @PostMapping("/create")
    public ResponseEntity<ApiResponse<FixedDeposit>> createFD(
            @AuthenticationPrincipal User user,
            @RequestBody Map<String, Object> payload) {

        String accountId = (String) payload.get("sourceAccountId");
        if (accountId == null) {
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error("sourceAccountId is required"));
        }

        Object amtRaw = payload.get("principalAmount");
        Object rateRaw = payload.get("interestRate");
        Object tenureRaw = payload.get("tenureMonths");
        Object autoRenewRaw = payload.get("autoRenew");

        if (amtRaw == null || tenureRaw == null) {
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error("principalAmount and tenureMonths are required"));
        }

        FixedDeposit fd = new FixedDeposit();
        fd.setPrincipalAmount(new BigDecimal(amtRaw.toString()));
        fd.setTenureMonths(Integer.parseInt(tenureRaw.toString()));

        // interest rate: use provided or calculate by tenure
        if (rateRaw != null) {
            fd.setInterestRate(new BigDecimal(rateRaw.toString()));
        } else {
            int tenure = fd.getTenureMonths();
            BigDecimal rate = tenure >= 36 ? new BigDecimal("7.25")
                    : tenure >= 24 ? new BigDecimal("7.00")
                    : tenure >= 12 ? new BigDecimal("6.50")
                    : new BigDecimal("5.50");
            fd.setInterestRate(rate);
        }

        fd.setAutoRenew(autoRenewRaw != null && Boolean.parseBoolean(autoRenewRaw.toString()));

        SavingsAccount sa = new SavingsAccount();
        sa.setId(UUID.fromString(accountId));
        fd.setSourceAccount(sa);

        FixedDeposit created = fixedDepositService.createFD(user.getId(), fd);
        return ResponseEntity.ok(ApiResponse.ok("Fixed Deposit created", created));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<FixedDeposit>>> getFDs(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(ApiResponse.ok(fixedDepositService.getFDs(user.getId())));
    }

    @PostMapping("/{id}/break")
    public ResponseEntity<ApiResponse<FixedDeposit>> breakFD(
            @AuthenticationPrincipal User user,
            @PathVariable UUID id) {
        FixedDeposit fd = fixedDepositService.breakFD(id, user.getId());
        return ResponseEntity.ok(ApiResponse.ok("Fixed Deposit broken, amount credited to account", fd));
    }
}

File: backend-python/com/gxs/bank/controller/FixedDepositController.py
from decimal import Decimal

from fastapi import APIRouter, Depends

from com.gxs.bank.config.database import get_db
from com.gxs.bank.config.serialization import serialize
from com.gxs.bank.dto.response.ApiResponse import ok
from com.gxs.bank.exception.BadRequestException import BadRequestException
from com.gxs.bank.model.FixedDeposit import FixedDeposit
from com.gxs.bank.security.JwtAuthenticationFilter import get_current_user
from com.gxs.bank.service.FixedDepositService import FixedDepositService


router = APIRouter(prefix="/api/fd", tags=["FixedDeposit"])


@router.post("/create")
def createFD(payload: dict, current_user=Depends(get_current_user), db=Depends(get_db)):
    accountId = payload.get("sourceAccountId")
    if accountId is None:
        raise BadRequestException("sourceAccountId is required")

    amount_raw = payload.get("principalAmount")
    rate_raw = payload.get("interestRate")
    tenure_raw = payload.get("tenureMonths")
    auto_renew_raw = payload.get("autoRenew")

    if amount_raw is None or tenure_raw is None:
        raise BadRequestException("principalAmount and tenureMonths are required")

    fd = FixedDeposit()
    fd.principalAmount = Decimal(str(amount_raw))
    fd.tenureMonths = int(str(tenure_raw))

    if rate_raw is not None:
        fd.interestRate = Decimal(str(rate_raw))
    else:
        tenure = fd.tenureMonths
        if tenure >= 36:
            fd.interestRate = Decimal("7.25")
        elif tenure >= 24:
            fd.interestRate = Decimal("7.00")
        elif tenure >= 12:
            fd.interestRate = Decimal("6.50")
        else:
            fd.interestRate = Decimal("5.50")

    if auto_renew_raw is None:
        fd.autoRenew = False
    elif isinstance(auto_renew_raw, bool):
        fd.autoRenew = auto_renew_raw
    else:
        fd.autoRenew = str(auto_renew_raw).strip().lower() == "true"
    fd.sourceAccountId = accountId

    created = FixedDepositService(db).createFD(current_user.id, fd)
    return ok(serialize(created), "Fixed Deposit created")


@router.get("")
def getFDs(current_user=Depends(get_current_user), db=Depends(get_db)):
    fds = FixedDepositService(db).getFDs(current_user.id)
    return ok(serialize(fds))


@router.post("/{fd_id}/break")
def breakFD(fd_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    fd = FixedDepositService(db).breakFD(fd_id, current_user.id)
    return ok(serialize(fd), "Fixed Deposit broken, amount credited to account")
